backends/ubpf/ubpfDeparser.cpp

```diff
@@ -0,0 +1,235 @@
+/*
+Copyright 2019 Orange
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "ubpfDeparser.h"
+#include "ubpfType.h"
+#include "frontends/p4/methodInstance.h"
+
+namespace UBPF {
+
+    UBPFDeparserTranslationVisitor::UBPFDeparserTranslationVisitor(
+            const UBPFDeparser *deparser) :
+            CodeGenInspector(deparser->program->refMap,
+                             deparser->program->typeMap), deparser(deparser),
+            p4lib(P4::P4CoreLibrary::instance) {
+        setName("UBPFDeparserTranslationVisitor");
+    }
+
+    void UBPFDeparserTranslationVisitor::compileEmitField(const IR::Expression *expr,
+                                                          cstring field,
+                                                          unsigned alignment,
+                                                          EBPF::EBPFType *type) {
+
+        unsigned widthToEmit = dynamic_cast<EBPF::IHasWidth *>(type)->widthInBits();
+        unsigned loadSize;
+        cstring swap = "";
+        if (widthToEmit <= 8) {
+            loadSize = 8;
+        } else if (widthToEmit <= 16) {
+            swap = "bpf_htons";
+            loadSize = 16;
+        } else if (widthToEmit <= 32) {
+            swap = "htonl";
+            loadSize = 32;
+        } else if (widthToEmit <= 64) {
+            swap = "htonll";
+            loadSize = 64;
```

Comment:
The ebpf compiler actually has a bug in the handling of htons; it is not clear that fields that do not have a size a power of 2 should be transformed using these macros. For example, the ebpf compiler incorrectly handles ethernet addresses. But perhaps you have tested this stuff. It is not very clear what header fields should be converted to machine order; I think that the ideal case would be not convert to machine order anything on parsing, but only when a field is used in some arithmetic operation.

Reply:
That's certainly a place for further optimizations. In comparison to ebpf compiler we have fixed the issue with Ethernet address (48 bit, < 64), IPv6 addresses (128 bit, > 64) and also the problem with writing a custom-width fields using `write_partial`, which didn't work properly in `ebpf` compiler.

Reply:
That's great. If you could at some point send a PR to fix that back-end it would be great.